compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/ModuleStopMethodGen.java

```diff
@@ -92,121 +89,56 @@
     public static final String ARR_VAR = "arrVar";
-    private final SymbolTable symbolTable;
     private final BIRVarToJVMIndexMap indexMap;
     private final JvmTypeGen jvmTypeGen;
     private final String strandMetadataClass;
 
-    public ModuleStopMethodGen(SymbolTable symbolTable, JvmTypeGen jvmTypeGen, JvmConstantsGen jvmConstantsGen) {
-        this.symbolTable = symbolTable;
+    public ModuleStopMethodGen(JvmTypeGen jvmTypeGen, JvmConstantsGen jvmConstantsGen) {
         indexMap = new BIRVarToJVMIndexMap(1);
         this.jvmTypeGen = jvmTypeGen;
         this.strandMetadataClass = jvmConstantsGen.getStrandMetadataConstantsClass();
     }
 
     public void generateExecutionStopMethod(ClassWriter cw, String initClass, BIRNode.BIRPackage module,
-                                            List<PackageID> imprtMods, AsyncDataCollector asyncDataCollector) {
+                                            AsyncDataCollector asyncDataCollector, Set<PackageID> immediateImports) {
         MethodVisitor mv = cw.visitMethod(Opcodes.ACC_PUBLIC + ACC_STATIC, MODULE_STOP_METHOD,
-                INIT_RUNTIME_REGISTRY, null, null);
+                MODULE_STOP_METHOD_DESC, null, null);
         mv.visitCode();
-
-        int schedulerIndex = indexMap.addIfNotExists(SCHEDULER_VAR, symbolTable.anyType);
-        // Create a scheduler. A new scheduler is used here, to make the stop function to not to
-        // depend/wait on whatever is being running on the background. eg: a busy loop in the main.
-        mv.visitTypeInsn(NEW, SCHEDULER);
-        mv.visitInsn(DUP);
-        mv.visitInsn(ICONST_1);
-        mv.visitInsn(ICONST_0);
-        mv.visitMethodInsn(INVOKESPECIAL, SCHEDULER, JVM_INIT_METHOD, "(IZ)V", false);
-        mv.visitVarInsn(ASTORE, schedulerIndex);
-
         String moduleInitClass = getModuleInitClassName(module.packageID);
         String fullFuncName = MethodGenUtils.calculateLambdaStopFuncName(module.packageID);
-        String lambdaName = generateStopDynamicLambdaBody(cw, initClass);
-        indexMap.addIfNotExists(FUTURE_VAR, symbolTable.anyType);
+        mv.visitFieldInsn(GETSTATIC, initClass, NO_OF_DEPENDANT_MODULES, "I");
+        mv.visitInsn(ICONST_1);
+        mv.visitInsn(ISUB);
+        mv.visitFieldInsn(PUTSTATIC, initClass, NO_OF_DEPENDANT_MODULES, "I");
+
+        mv.visitFieldInsn(GETSTATIC, initClass, NO_OF_DEPENDANT_MODULES, "I");
+        Label labelIf = new Label();
+        mv.visitJumpInsn(IFLE, labelIf);
+        mv.visitInsn(RETURN);
+        mv.visitLabel(labelIf);
+
         scheduleStopLambda(mv, initClass, fullFuncName, moduleInitClass, asyncDataCollector);
-        int i = imprtMods.size() - 1;
-        while (i >= 0) {
-            PackageID id = imprtMods.get(i);
-            i -= 1;
-            fullFuncName = MethodGenUtils.calculateLambdaStopFuncName(id);
+
+        Iterator<PackageID> immediateImportsIterator = (new LinkedList<>(immediateImports)).descendingIterator();
```

Comment:
Why we have removed the while loop to loop through decendingOrder and create another list and use iterator?

Reply:
I rechecked the logic and seems like the moduleStop() invocation is not needed to be in any particular order. Since we have validation to check whether there are any modules using a given module before invoking its moduleStop(), the order won't matter. I will remove the `descendingOrder` iterator and refactor the logic.

Please refer to [this document](https://docs.google.com/document/d/1uSiPY75GD0qnDH1GRnXM6vdpegDmgwe72mOW4Hoty2o/edit?usp=sharing) for further details about moduleStop() order.

Reply:
Removed the unnecessary iteration logic in 8da28c04c1397a39613ef773b0f56c5989ce3016.

Reply:
Turned out the test failure was due to a change in the `moduleStop()` order of two sibling modules. The `moduleStop()` order of sibling modules does not affect the final execution. 

Test is passing now after updating the order in bc2510a24dc18265d94cfb0ec78159b3f413cefa.